pkg/resourcegroup/controllers/typeresolver/typeresolver.go

```diff
@@ -16,39 +16,47 @@ package typeresolver
 
 import (
 	"context"
+	"errors"
+	"fmt"
 	"sync"
 
 	"github.com/go-logr/logr"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/client-go/discovery"
 	ctrl "sigs.k8s.io/controller-runtime"
-	"sigs.k8s.io/controller-runtime/pkg/controller"
-	"sigs.k8s.io/controller-runtime/pkg/handler"
-	"sigs.k8s.io/controller-runtime/pkg/reconcile"
-	"sigs.k8s.io/controller-runtime/pkg/source"
 )
 
 // TypeResolver keeps the preferred GroupVersionKind for all the
 // types in the cluster.
 type TypeResolver struct {
 	log         logr.Logger
 	mu          sync.Mutex
-	dc          *discovery.DiscoveryClient
+	dc          discovery.DiscoveryInterface
 	typeMapping map[schema.GroupKind]schema.GroupVersionKind
 }
 
 // Refresh refreshes the type mapping by querying the api server
-func (r *TypeResolver) Refresh() {
+func (r *TypeResolver) Refresh() error {
 	mapping := make(map[schema.GroupKind]schema.GroupVersionKind)
 	apiResourcesList, err := discovery.ServerPreferredResources(r.dc)
 	if err != nil {
-		r.log.Error(err, "Unable to fetch api resources list by dynamic client")
-		return
+		if isStaleGroupDiscoveryError(err) {
+			// Log and continue, using the cached APIs.
+			// Stale groups means one or more of the APIService backends are
+			// unhealthy. This is especially common for metrics providers.
+			// If the API provider of a managed resource remains unhealthy, it
+			// will be surfaced as a resource-specific API error later, when
+			// watched or listed, which are retried and won't prevent updating
+			// the status of other managed resources.
+			r.log.Error(err, "Failed to discover API resources; using cached results")
+		} else {
```

Comment:
Will the `err` contain the stale gv for reference?

Reply:
yup https://github.com/kubernetes/kubernetes/blob/release-1.31/staging/src/k8s.io/client-go/discovery/discovery_client.go#L437